File: assignment1_rt/src/ui_node1.cpp
#include "ros/ros.h"
#include "turtlesim/Spawn.h"
#include "geometry_msgs/Twist.h"
#include <iostream>
#include <string>

// Function to spawn a turtle
void spawnTurtle(const std::string &name, float x, float y, float theta) {
    ros::NodeHandle nh;
    ros::ServiceClient client = nh.serviceClient<turtlesim::Spawn>("/spawn");

    turtlesim::Spawn srv;
    srv.request.x = x;
    srv.request.y = y;
    srv.request.theta = theta;
    srv.request.name = name;

    if (client.call(srv)) {
        ROS_INFO("Spawned turtle '%s' at (%f, %f, %f)", name.c_str(), x, y, theta);
    } else {
        ROS_ERROR("Failed to call service /spawn");
        ros::shutdown();
    }
}

// Function to get user input for turtle selection and velocities
bool getUserInput(std::string &turtle, float &linear, float &angular) {
    int choice;
    std::cout << "Which turtle to control? (1 for turtle1, 2 for turtle2): ";
    std::cin >> choice;

    if (choice == 1) {
        turtle = "turtle1";
    } else if (choice == 2) {
        turtle = "turtle2";
    } else {
        std::cerr << "Invalid choice. Please enter 1 or 2." << std::endl;
        return false;
    }

    std::cout << "Enter linear velocity: ";
    if (!(std::cin >> linear)) {
        std::cerr << "Invalid input. Please enter numeric values for velocity." << std::endl;
        return false;
    }

    std::cout << "Enter angular velocity: ";
    if (!(std::cin >> angular)) {
        std::cerr << "Invalid input. Please enter numeric values for velocity." << std::endl;
        return false;
    }

    return true;
}

// Function to send velocity commands to the turtle
void sendCommand(ros::Publisher &pub, float linear, float angular) {
    geometry_msgs::Twist cmd;
    cmd.linear.x = linear;
    cmd.angular.z = angular;

    pub.publish(cmd);
    ROS_INFO("Command sent: linear=%f, angular=%f", linear, angular);

    ros::Duration(1.0).sleep(); // Send the command for 1 second

    cmd.linear.x = 0.0;
    cmd.angular.z = 0.0;
    pub.publish(cmd);
    ROS_INFO("Turtle stopped.");
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "ui_node");

    ros::NodeHandle nh;

    // Spawn a second turtle
    spawnTurtle("turtle2", 3.0, 3.0, 0.0);

    // Publishers for turtle1 and turtle2
    ros::Publisher pub_turtle1 = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
    ros::Publisher pub_turtle2 = nh.advertise<geometry_msgs::Twist>("/turtle2/cmd_vel", 10);

    ROS_INFO("User interface is ready. Enter commands:");

    while (ros::ok()) {
        try {
            std::string turtle;
            float linear, angular;

            if (!getUserInput(turtle, linear, angular)) {
                continue;
            }

            if (turtle == "turtle1") {
                sendCommand(pub_turtle1, linear, angular);
            } else if (turtle == "turtle2") {
                sendCommand(pub_turtle2, linear, angular);
            }

        } catch (const std::exception &e) {
            ROS_ERROR("Exception occurred: %s", e.what());
            break;
        }
    }

    return 0;
}

File: assignment1_rt/scripts/ui_node1.py
#!/usr/bin/env python3
"""Provides a ROS node for spawning and controlling turtles in turtlesim."""

import rospy
from turtlesim.srv import Spawn, SpawnRequest
from geometry_msgs.msg import Twist
import sys
from typing import Optional, Tuple


def spawn_turtle(name: str, x: float, y: float, theta: float) -> None:
    """Call the spawn service to create a new turtle.

    Args:
        name (str): Name of the turtle to spawn.
        x (float): X-coordinate of the spawn location.
        y (float): Y-coordinate of the spawn location.
        theta (float): Initial orientation of the turtle in radians.

    Raises:
        SystemExit: If the spawn service call fails.

    """
    rospy.wait_for_service("/spawn")
    try:
        spawn_service = rospy.ServiceProxy("/spawn", Spawn)
        req = SpawnRequest()
        req.x = x
        req.y = y
        req.theta = theta
        req.name = name
        spawn_service(req)
        rospy.loginfo(f"Spawned {name} at ({x}, {y}, {theta})")
    except rospy.ServiceException as e:
        rospy.logerr(f"Failed to spawn {name}: {e}")
        sys.exit(1)


def get_user_input() -> Tuple[Optional[str], Optional[float], Optional[float]]:
    """Get user input for turtle selection and velocities.

    Returns:
        tuple: A tuple containing:
            - (str): The selected turtle name ("turtle1" or "turtle2").
            - (float): The linear velocity.
            - (float): The angular velocity.
            
        Returns (None, None, None) if the input is invalid.
    """
    choice = input("Which turtle to control? (1 for turtle1, 2 for turtle2): ")
    if choice == "1":
        turtle = "turtle1"
    elif choice == "2":
        turtle = "turtle2"
    else:
        print("Invalid choice. Please enter 1 or 2.")
        return None, None, None

    try:
        vel = float(input("Enter linear velocity: "))
        ang = float(input("Enter angular velocity: "))
    except ValueError:
        print("Invalid input. Please enter numeric values for velocity.")
        return None, None, None

    return turtle, vel, ang


def send_command(pub: rospy.Publisher, vel: float, ang: float) -> None:
    """Send velocity command to the turtle.

    Args:
        pub (rospy.Publisher): The publisher object for sending commands to the turtle.
        vel (float): Linear velocity.
        ang (float): Angular velocity.

    """
    cmd = Twist()
    cmd.linear.x = vel
    cmd.angular.z = ang
    pub.publish(cmd)
    rospy.loginfo(f"Command sent: linear={vel}, angular={ang}")
    rospy.sleep(1)
    cmd.linear.x = 0.0
    cmd.angular.z = 0.0
    pub.publish(cmd)
    rospy.loginfo("Turtle stopped.")


def main() -> None:
    """Initialize the ROS node and handle user input for controlling turtles.

    This function sets up the ROS node, spawns a second turtle, and continuously
    accepts user input to control the movement of either turtle1 or turtle2.
    """
    rospy.init_node("ui_node", anonymous=True)

    # Turtle spawning
    spawn_turtle("turtle2", 3.0, 3.0, 0.0)

    # Creating Publishers
    pub_turtle1 = rospy.Publisher("/turtle1/cmd_vel", Twist, queue_size=10)
    pub_turtle2 = rospy.Publisher("/turtle2/cmd_vel", Twist, queue_size=10)

    rospy.loginfo("User interface is ready. Enter commands:")

    while not rospy.is_shutdown():
        try:
            turtle, vel, ang = get_user_input()
            if turtle is None:
                continue

            if turtle == "turtle1":
                send_command(pub_turtle1, vel, ang)
            elif turtle == "turtle2":
                send_command(pub_turtle2, vel, ang)

        except rospy.ROSInterruptException:
            rospy.loginfo("Shutting down UI node.")
            break
        except KeyboardInterrupt:
            print("Exiting...")
            break


if __name__ == "__main__":
    main()

